src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypeDiagnosticAnalyzer.cs

```diff
@@ -107,9 +107,18 @@ protected override bool TryAnalyzeVariableDeclaration(TypeSyntax typeName, Seman
                 }
 
                 var variable = variableDeclaration.Variables.Single();
-                if (AssignmentSupportsStylePreference(
-                        variable.Identifier, typeName, variable.Initializer.Value,
-                        semanticModel, optionSet, cancellationToken))
+                var initializer = variable.Initializer.Value;
+
+                // Discourage using "var pointer = stackalloc", after introducing C# 7.2, as it might be confusing to users expecting a Span<>
+                // Check descendant nodes as well, because Span-creating stackallocs can be nested in other nodes (like conditional operators and casts).
+                // https://github.com/dotnet/roslyn/issues/22768
+                if (initializer.DescendantNodesAndSelf().Any(node => node.IsKind(SyntaxKind.StackAllocArrayCreationExpression)))
```

Comment:
❓ Why is this needed? We already have the following block which is intended to catch exactly this case. For example, no special casing is needed for `IFormattable f = $"";`.

https://github.com/dotnet/roslyn/blob/4b1c677e0161df22f12073a1493747532152b115/src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypeDiagnosticAnalyzer.cs#L180-L184

Reply:
Because in cases where compiler can bind the node successfully without errors, it will be replaced with a `BoundConvertedStackAllocExpression`. That's what the semantic model will return.